cmd/fips/main.go

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"log"
+	"os"
+	"strings"
+
+	"golang.org/x/tools/go/packages"
+)
+
+type fnReport struct {
+	PackageID        string
+	Name             string
+	HasBoringEnabled bool
+	BoringCalls      []string
+}
+
+const description = `
+Fips examines Go source code and reports which public APIs are backed by openssl
+primitives, together with other usefull information such as the name of the
+openssl function names.
+
+Set the appropiate GOROOT when checking std packagesnot in your default Go toolchain. 
+
+Packages are specified using the notation of "go list".
+This command checks the package in the current directory:
+
+	go run ./cmd/fips
+
+whereas this one checks the packages whose path is provided:
+
+	go run ./cmd/fips crypto/...
+`
+
+func main() {
+	log.SetFlags(0)
+	log.SetPrefix("fips: ")
+
+	flag.Usage = func() {
+		flag.PrintDefaults()
+		fmt.Fprintf(flag.CommandLine.Output(), "%s\n\n", description)
+	}
+
+	flag.Parse()
+
+	pkgs, err := parsePackages()
+	if err != nil {
+		log.Fatalf("load: %v\n", err)
+		os.Exit(1)
+	}
+	if packages.PrintErrors(pkgs) > 0 {
+		os.Exit(1)
+	}
+
+	for _, pkg := range pkgs {
+		// We are only interested in public packages.
+		if strings.Contains(pkg.ID, "/internal/") {
+			continue
+		}
+		for _, src := range pkg.Syntax {
+			for _, decl := range src.Decls {
+				if decl, ok := decl.(*ast.FuncDecl); ok {
+					if !ast.IsExported(decl.Name.Name) {
+						// We are only interested in exported declarations.
+						continue
+					}
+					var report fnReport
+					report.PackageID = pkg.ID
+					report.Name = decl.Name.Name
+					if processFuncDecl(decl, pkg.Syntax, &report) {
+						log.Println(report)
+					}
+				}
+			}
+		}
+	}
+}
+
+func parsePackages() ([]*packages.Package, error) {
+	cfg := &packages.Config{
+		Env: append(os.Environ(), "GOOS=linux"),
```

Comment:
This is a tricky one. You cannot set `GOOS=linux` on Windows to check for Linux APIs (and vice-versa) when running the command using `go run`, as it will build the binary for the OS specified in `GOOS`, which in this situation is not to host OS so it won't be executable.

I let it like a hardcoded env variable until I find a better solution, such as passing the GOOS as a cmd argument (open to ideas). But yes, it deserve a comment!